File: clients/php/src/CheFuAcademyClient.php
<?php

declare(strict_types=1);

namespace CheFu\Academy;

final class CheFuAcademyClient
{
    public const DEFAULT_BASE_URL = 'https://api.chefuinc.com/api';

    public function __construct(
        private ?string $apiKey = null,
        private ?string $authToken = null,
        private string $baseUrl = self::DEFAULT_BASE_URL,
        private int $timeoutSeconds = 10,
    ) {
        $this->baseUrl = rtrim($this->baseUrl, '/');
    }

    public function setAuthToken(string $token): void
    {
        $this->authToken = $token;
    }

    /** @return array<string, mixed> */
    public function login(string $email, string $password): array
    {
        $session = $this->request('POST', '/auth/login', body: [
            'email' => $email,
            'password' => $password,
        ], apiKeyAuth: false);
        $token = $session['idToken'] ?? $session['token'] ?? null;
        if (is_string($token) && $token !== '') {
            $this->setAuthToken($token);
        }
        return $session;
    }

    /** @return array<string, mixed> */
    public function register(string $email, string $password, string $fullname): array
    {
        return $this->request('POST', '/auth/register', body: [
            'email' => $email,
            'password' => $password,
            'fullname' => $fullname,
        ], apiKeyAuth: false);
    }

    /** @return array<string, mixed> */
    public function refresh(string $refreshToken): array
    {
        $session = $this->request('POST', '/auth/refresh', body: [
            'refreshToken' => $refreshToken,
        ], apiKeyAuth: false);
        $token = $session['idToken'] ?? $session['token'] ?? null;
        if (is_string($token) && $token !== '') {
            $this->setAuthToken($token);
        }
        return $session;
    }

    /** @return array<string, mixed> */
    public function listCourses(array $query = []): array
    {
        return $this->request('GET', '/courses', query: $query);
    }

    /** @return array<string, mixed> */
    public function searchCourses(array $query = []): array
    {
        return $this->request('GET', '/courses/search', query: $query);
    }

    /** @return array<string, mixed> */
    public function featuredCourses(array $query = []): array
    {
        return $this->request('GET', '/courses/featured', query: $query);
    }

    /** @return array<string, mixed> */
    public function courseCategories(): array
    {
        return $this->request('GET', '/courses/categories');
    }

    /** @return array<string, mixed> */
    public function course(string $courseId): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId));
    }

    /** @return array<string, mixed> */
    public function courseChapters(string $courseId): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/chapters');
    }

    /** @return array<string, mixed> */
    public function courseChapter(string $courseId, int $chapterIndex): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/chapters/' . $chapterIndex);
    }

    /** @return array<string, mixed> */
    public function courseLessons(string $courseId, int $chapterIndex): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/chapters/' . $chapterIndex . '/lessons');
    }

    /** @return array<string, mixed> */
    public function courseQuiz(string $courseId): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/quiz');
    }

    /** @return array<string, mixed> */
    public function courseFlashcards(string $courseId): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/flashcards');
    }

    /** @return array<string, mixed> */
    public function courseQa(string $courseId): array
    {
        return $this->request('GET', '/courses/' . rawurlencode($courseId) . '/qa');
    }

    /** @return array<string, mixed> */
    public function listVideos(array $query = []): array
    {
        return $this->request('GET', '/videos', query: $query);
    }

    /** @return array<string, mixed> */
    public function searchVideos(array $query = []): array
    {
        return $this->request('GET', '/videos/search', query: $query);
    }

    /** @return array<string, mixed> */
    public function videosByCategory(string $category): array
    {
        return $this->request('GET', '/videos/category/' . rawurlencode($category));
    }

    /** @return array<string, mixed> */
    public function video(string $videoId): array
    {
        return $this->request('GET', '/videos/' . rawurlencode($videoId));
    }

    /** @return array<string, mixed> */
    public function createKey(?string $name = null): array
    {
        return $this->request('POST', '/keys/create', body: ['name' => $name], userAuth: true, apiKeyAuth: false);
    }

    /** @return array<int, array<string, mixed>> */
    public function listKeys(): array
    {
        return $this->request('GET', '/keys/list', userAuth: true, apiKeyAuth: false);
    }

    /** @return array<string, mixed> */
    public function revokeKey(string $keyId): array
    {
        return $this->request('POST', '/keys/revoke', body: ['keyId' => $keyId], userAuth: true, apiKeyAuth: false);
    }

    /**
     * @param array<string, mixed> $query
     * @param array<string, mixed>|null $body
     * @return array<mixed>
     */
    private function request(
        string $method,
        string $path,
        array $query = [],
        ?array $body = null,
        bool $userAuth = false,
        bool $apiKeyAuth = true,
    ): array {
        $token = $userAuth ? $this->authToken : $this->apiKey;
        if ($userAuth && !$token) {
            throw new CheFuAcademyException('User authentication is required.', 401);
        }
        if ($apiKeyAuth && !$userAuth && !$token) {
            throw new CheFuAcademyException('API key is required.', 401);
        }

        $cleanQuery = array_filter($query, static fn ($value): bool => $value !== null && $value !== '');
        $url = $this->baseUrl . $path . ($cleanQuery ? '?' . http_build_query($cleanQuery) : '');
        $headers = ['Accept: application/json'];
        if ($token) {
            $headers[] = 'Authorization: Bearer ' . $token;
        }

        $curl = curl_init($url);
        curl_setopt_array($curl, [
            CURLOPT_CUSTOMREQUEST => $method,
            CURLOPT_RETURNTRANSFER => true,
            CURLOPT_TIMEOUT => $this->timeoutSeconds,
            CURLOPT_HTTPHEADER => $headers,
        ]);

        if ($body !== null) {
            $headers[] = 'Content-Type: application/json';
            curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
            curl_setopt($curl, CURLOPT_POSTFIELDS, json_encode($body, JSON_THROW_ON_ERROR));
        }

        $responseBody = curl_exec($curl);
        $statusCode = curl_getinfo($curl, CURLINFO_RESPONSE_CODE);
        $error = curl_error($curl);
        curl_close($curl);

        if ($responseBody === false) {
            throw new CheFuAcademyException('Network error: ' . $error);
        }

        $decoded = $responseBody === '' ? [] : json_decode($responseBody, true);
        if (!is_array($decoded)) {
            $decoded = [];
        }

        if ($statusCode < 200 || $statusCode >= 300) {
            throw new CheFuAcademyException($this->errorMessage($decoded, $responseBody), $statusCode);
        }

        return $decoded;
    }

    /** @param array<mixed> $decoded */
    private function errorMessage(array $decoded, string $raw): string
    {
        $message = $decoded['message'] ?? null;
        if (is_array($message)) {
            return implode(' ', array_map('strval', $message));
        }
        if (is_string($message) && $message !== '') {
            return $message;
        }
        $error = $decoded['error'] ?? null;
        if (is_string($error) && $error !== '') {
            return $error;
        }
        return $raw !== '' ? $raw : 'CheFu Academy request failed.';
    }
}

File: clients/java/src/main/java/com/chefu/academy/CheFuAcademyClient.java
package com.chefu.academy;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.StringJoiner;

public final class CheFuAcademyClient {
    public static final String DEFAULT_BASE_URL = "https://api.chefuinc.com/api";

    private final ObjectMapper mapper = new ObjectMapper();
    private final HttpClient httpClient;
    private final String baseUrl;
    private String apiKey;
    private String authToken;

    public CheFuAcademyClient(String apiKey, String authToken, String baseUrl) {
        this.apiKey = apiKey;
        this.authToken = authToken;
        this.baseUrl = trimTrailingSlash(baseUrl == null || baseUrl.isBlank() ? DEFAULT_BASE_URL : baseUrl);
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();
    }

    public static CheFuAcademyClient withApiKey(String apiKey) {
        return new CheFuAcademyClient(apiKey, null, DEFAULT_BASE_URL);
    }

    public void setAuthToken(String token) {
        this.authToken = token;
    }

    public JsonNode login(String email, String password) {
        ObjectNode body = mapper.createObjectNode()
            .put("email", email)
            .put("password", password);
        JsonNode session = request("POST", "/auth/login", null, body, false, false);
        String token = text(session, "idToken");
        if (token == null || token.isBlank()) {
            token = text(session, "token");
        }
        if (token != null && !token.isBlank()) {
            setAuthToken(token);
        }
        return session;
    }

    public JsonNode register(String email, String password, String fullname) {
        ObjectNode body = mapper.createObjectNode()
            .put("email", email)
            .put("password", password)
            .put("fullname", fullname);
        return request("POST", "/auth/register", null, body, false, false);
    }

    public JsonNode refresh(String refreshToken) {
        ObjectNode body = mapper.createObjectNode().put("refreshToken", refreshToken);
        JsonNode session = request("POST", "/auth/refresh", null, body, false, false);
        String token = text(session, "idToken");
        if (token == null || token.isBlank()) {
            token = text(session, "token");
        }
        if (token != null && !token.isBlank()) {
            setAuthToken(token);
        }
        return session;
    }

    public JsonNode listCourses(Map<String, ?> query) {
        return request("GET", "/courses", query, null, false, true);
    }

    public JsonNode searchCourses(Map<String, ?> query) {
        return request("GET", "/courses/search", query, null, false, true);
    }

    public JsonNode featuredCourses(Map<String, ?> query) {
        return request("GET", "/courses/featured", query, null, false, true);
    }

    public JsonNode courseCategories() {
        return request("GET", "/courses/categories", null, null, false, true);
    }

    public JsonNode course(String courseId) {
        return request("GET", "/courses/" + encode(courseId), null, null, false, true);
    }

    public JsonNode courseChapters(String courseId) {
        return request("GET", "/courses/" + encode(courseId) + "/chapters", null, null, false, true);
    }

    public JsonNode courseChapter(String courseId, int chapterIndex) {
        return request("GET", "/courses/" + encode(courseId) + "/chapters/" + chapterIndex, null, null, false, true);
    }

    public JsonNode courseLessons(String courseId, int chapterIndex) {
        return request("GET", "/courses/" + encode(courseId) + "/chapters/" + chapterIndex + "/lessons", null, null, false, true);
    }

    public JsonNode courseQuiz(String courseId) {
        return request("GET", "/courses/" + encode(courseId) + "/quiz", null, null, false, true);
    }

    public JsonNode courseFlashcards(String courseId) {
        return request("GET", "/courses/" + encode(courseId) + "/flashcards", null, null, false, true);
    }

    public JsonNode courseQa(String courseId) {
        return request("GET", "/courses/" + encode(courseId) + "/qa", null, null, false, true);
    }

    public JsonNode listVideos(Map<String, ?> query) {
        return request("GET", "/videos", query, null, false, true);
    }

    public JsonNode searchVideos(Map<String, ?> query) {
        return request("GET", "/videos/search", query, null, false, true);
    }

    public JsonNode videosByCategory(String category) {
        return request("GET", "/videos/category/" + encode(category), null, null, false, true);
    }

    public JsonNode video(String videoId) {
        return request("GET", "/videos/" + encode(videoId), null, null, false, true);
    }

    public JsonNode createKey(String name) {
        ObjectNode body = mapper.createObjectNode().put("name", name);
        return request("POST", "/keys/create", null, body, true, false);
    }

    public JsonNode listKeys() {
        return request("GET", "/keys/list", null, null, true, false);
    }

    public JsonNode revokeKey(String keyId) {
        ObjectNode body = mapper.createObjectNode().put("keyId", keyId);
        return request("POST", "/keys/revoke", null, body, true, false);
    }

    private JsonNode request(
        String method,
        String path,
        Map<String, ?> query,
        JsonNode body,
        boolean userAuth,
        boolean apiKeyAuth
    ) {
        String token = userAuth ? authToken : apiKey;
        if (userAuth && (token == null || token.isBlank())) {
            throw new CheFuAcademyException("User authentication is required.", 401);
        }
        if (apiKeyAuth && !userAuth && (token == null || token.isBlank())) {
            throw new CheFuAcademyException("API key is required.", 401);
        }

        String requestUrl = baseUrl + path + queryString(query);
        HttpRequest.Builder builder = HttpRequest.newBuilder()
            .uri(URI.create(requestUrl))
            .timeout(Duration.ofSeconds(10))
            .header("Accept", "application/json");
        if (token != null && !token.isBlank()) {
            builder.header("Authorization", "Bearer " + token);
        }
        if (body != null) {
            builder.header("Content-Type", "application/json");
            builder.method(method, HttpRequest.BodyPublishers.ofString(body.toString()));
        } else {
            builder.method(method, HttpRequest.BodyPublishers.noBody());
        }

        try {
            HttpResponse<String> response = httpClient.send(builder.build(), HttpResponse.BodyHandlers.ofString());
            if (response.statusCode() < 200 || response.statusCode() >= 300) {
                throw new CheFuAcademyException(errorMessage(response.body()), response.statusCode());
            }
            return response.body().isBlank() ? mapper.createObjectNode() : mapper.readTree(response.body());
        } catch (IOException exception) {
            throw new CheFuAcademyException("Network or JSON error: " + exception.getMessage());
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            throw new CheFuAcademyException("Request interrupted.");
        }
    }

    private String errorMessage(String body) {
        try {
            JsonNode node = mapper.readTree(body);
            JsonNode message = node.get("message");
            if (message != null) {
                if (message.isArray()) {
                    StringJoiner joiner = new StringJoiner(" ");
                    message.forEach(item -> joiner.add(item.asText()));
                    return joiner.toString();
                }
                return message.asText();
            }
            JsonNode error = node.get("error");
            if (error != null) {
                return error.asText();
            }
        } catch (IOException ignored) {
            return body;
        }
        return body == null || body.isBlank() ? "CheFu Academy request failed." : body;
    }

    private String queryString(Map<String, ?> query) {
        if (query == null || query.isEmpty()) {
            return "";
        }
        StringJoiner joiner = new StringJoiner("&", "?", "");
        query.forEach((key, value) -> {
            if (value != null && !value.toString().isBlank()) {
                joiner.add(encode(key) + "=" + encode(value.toString()));
            }
        });
        return joiner.length() <= 1 ? "" : joiner.toString();
    }

    private static String trimTrailingSlash(String value) {
        return value.replaceAll("/+$", "");
    }

    private static String encode(String value) {
        return URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String text(JsonNode node, String field) {
        JsonNode value = node.get(field);
        return value == null || value.isNull() ? null : value.asText();
    }
}
